### PR TITLE
1.19.0 support

### PR DESCRIPTION
**Pull request by extremeheat**
With chat signing support

**Todo**
- Verify logic is correct for offline mode

**Adds**
- new client.signMessage and client.verifyMessage utility methods
- binary concatenation function (could possibly fit  into protodef)

### CODE REVIEW

File: src/client/microsoftAuth.js [hunk 1]
```diff
@@ -34,6 +37,19 @@ async function authenticate (client, options) {
   }
   client.session = session
   client.username = profile.name
+  client.profileKeys = {}
```
Comment (extremeheat) [inline src/client/microsoftAuth.js:40]: Delete

File: src/server/login.js [hunk 2]
```diff
@@ -27,9 +29,39 @@ module.exports = function (client, server, options) {
...
+        if (!crypto.verify('RSA-SHA1', Buffer.from(signable, 'utf8'), mojangPubicKey, packet.signature.signature)) {
+          client.end('multiplayer.disconnect.invalid_public_key_signature')
+          return
+        }
+        client.profileKeys = { public: publicKey, publicPEM }
+      } catch (err) {
+        client.end('multiplayer.disconnect.invalid_public_key')
+        return
+      }
+    } else if (options.enforceSecureProfile) {
```
Comment (extremeheat) [inline src/server/login.js:60]: Dupe condition of above L38

File: src/client/microsoftAuth.js [hunk 3]
```diff
@@ -34,6 +37,19 @@ async function authenticate (client, options) {
...
+
+  if (certificates) {
+    client.profileKeys = {
+      publicPEM: certificates.keyPair.publicKey,
```
Comment (extremeheat) [inline src/client/microsoftAuth.js:44]: Maybe some of this transformation could be handled in prismarine-auth

Comment (Kashalls) [reply src/client/microsoftAuth.js:44]: Crypto should be done in its own class/util in nmp.

Comment (extremeheat) [reply src/client/microsoftAuth.js:44]: Will move this to pauth as it should not be returning unabstracted 1:1 web request data

**Comment by extremeheat** [comment]
Need to handle offline mode on server also. I'm assuming server disables chat signing for the client if the server is in offline mode, maybe through setting hasVerifyToken to true. Need to investigate that, cc @rob9315

**Comment by rom1504** [comment]
this looks almost ready now

**Comment by extremeheat** [comment]
Passing. Should be possible to add some sort of test for the chat signing, but it's a bit tricky because that would involve a bit of boilerplate and patching mojang public key constant.

**Comment by xtexChooser** [comment]
Is there any way to use the 1.19 support directly?

I knew this is unstable, but I need 1.19.

Thanks.

**Comment by vincss** [comment]
Hi !
When I try to login using this version from https://github.com/vincss/mcsleepingserverstarter, I got this error message : 
image
 And in my logs : 
Chunk size is 848 but only 831 was read ; partial packet : {"name":"login_start","params":{"username":"Vincouille","signature":{"timestamp":[386,-1950364232],"publicKey":{"type":"Buffer","data":[48,130,1,34,48,13,6,9,42,134,72,134
Hope it helps !

**Comment by extremeheat** [comment]
Are your dependencies up to date and are you using 1.19.0 (not 1.19.1) ? Sending an updated login packet allows my client to connect to an empty world

image

**Comment by vincss** [comment]
> @vincss
> 
> Are your dependencies up to date and are you using 1.19.0 (not 1.19.1) ? Sending an updated login packet allows my client to connect to an empty world
> 
> image

I was on a 1.19.2, but indeed in 1.19 it's working better. Thanks.

**Comment by rom1504** [comment]
how far is this from ready ?

**Comment by rom1504** [comment]
ok looked at the code

looks good to me

I would appreciate if a few more people could try to boot it up and check everything seems to work for them

**Comment by SpencerBurgess** [comment]
This doesn't update the README.md to include 1.19.0. Is that intentional or is that added with the next release PR?

**Comment by rom1504** [comment]
No, please add it

**Pull request merged**